File: python_kata/src/trees/binary_tree/post_order_traversal.py
from typing import List
from trees.binary_tree.node import BinaryNode


def walk(curr: BinaryNode[int] | None, path: List[int]) -> List[int]:
    if not curr:
        return path

    # Pre

    # Recursive
    walk(curr.left, path)
    walk(curr.right, path)

    # Post
    path.append(curr.val)
    return path

def post_order_traversal(head: BinaryNode[int]) -> List[int]:
    return walk(head, [])

File: typescript-kata/src/trees/binary-tree/postOrderTraversal.ts
import { BinaryNode, BinaryTree } from "./node";

function walk(curr: BinaryNode<number> | undefined, path: number[]): number[] {
  if (!curr) {
    return path;
  }

  // Pre
  //
  // Recursive
  walk(curr.left, path);
  walk(curr.right, path);

  // Post
  path.push(curr.val);
  return path;
}

export default function PostOrderTraversal(tree: BinaryTree<number>): number[] {
  if (!tree.root) {
    return [];
  }
  return walk(tree.root, []);
}
